core/src/libraries/Oracle.sol

```diff
@@ -86,6 +86,75 @@ library Oracle {
         }
     }
 
+    /**
+     * @notice Searches for oracle observations nearest to the `target` time. If `target` lies between two existing
+     * observations, linearly interpolate between them. If `target` is newer than the most recent observation,
+     * we interpolate between the most recent one and a hypothetical one taken at the current block.
+     * @dev As long as `target <= block.timestamp`, return values should match what you'd get from Uniswap:
+     * ```
+     *   uint32[] memory secondsAgos = new uint32[](1);
+     *   secondsAgos[0] = block.timestamp - target;
+     *   (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s) = pool.observe(
+     *     secondsAgos
+     *   );
+     * ```
+     * @param pool The Uniswap pool to examine
+     * @param target The timestamp of the desired observation
+     * @param seed The index of `pool.observations` where we start our search. Can be determined off-chain to make
+     * this method more efficient than Uniswap's binary search.
+     * @param tick The current tick (from pool.slot0())
+     * @param observationIndex The current observation index (from pool.slot0())
+     * @param observationCardinality The current observation cardinality (from pool.slot0())
+     * @param liquidity The current liquidity depth at `tick` (from pool.liquidity());
+     * @return tickCumulative The tick * time elapsed since `pool` was first initialized
+     * @return secondsPerLiquidityCumulativeX128 The time elapsed / max(1, liquidity) since `pool` was first initialized
+     */
+    function observe(
+        IUniswapV3Pool pool,
+        uint32 target,
+        uint256 seed,
+        int24 tick,
+        uint16 observationIndex,
+        uint16 observationCardinality,
+        uint128 liquidity
+    ) internal view returns (int56, uint160) {
+        unchecked {
+            seed %= observationCardinality;
+            (uint32 timeL, int56 tickCumL, uint160 liqCumL, ) = pool.observations(seed);
+
+            for (uint256 i = 0; i < observationCardinality; i++) {                
+                if (timeL == target) {
+                    return (tickCumL, liqCumL);
+                }
+
+                if (timeL < target && seed == observationIndex) {
+                    uint56 delta = uint56(target - timeL);
+                    return (
+                        tickCumL + tick * int56(delta),
+                        liqCumL + (uint160(delta) << 128) / (liquidity > 0 ? liquidity : 1)
```

Comment:
Why the `<<128`? I understand what you're doing in the line above, but this one I'm not so sure. Given that your `_comparative_` tests pass though, it must be right, but I don't see why you need this.

Reply:
The `liquidity` value from Uniswap has an extra factor of 2^128 in it, so when we're taking the reciprocal we need that same factor in the numerator.